cmake/app/boilerplate.cmake

```diff
@@ -26,42 +26,66 @@ cmake_policy(SET CMP0002 NEW)
 # CMP0079: "target_link_libraries() allows use with targets in other directories"
 cmake_policy(SET CMP0079 OLD)
 
-define_property(GLOBAL PROPERTY ZEPHYR_LIBS
-    BRIEF_DOCS "Global list of all Zephyr CMake libs that should be linked in"
-    FULL_DOCS  "Global list of all Zephyr CMake libs that should be linked in.
+get_property(IMAGE GLOBAL PROPERTY IMAGE)
+
+if(IMAGE)
+  set(FIRST_BOILERPLATE_EXECUTION 0)
+else()
+  set(FIRST_BOILERPLATE_EXECUTION 1)
+endif()
+
+if (NOT FIRST_BOILERPLATE_EXECUTION)
+  # Clear the Kconfig namespace of the other image.
```

Comment:
I am concerned about what the need for this loop means about this feature's effect on the build system's namespaces.

I don't have your deep knowledge of the build system, but this seems to imply at least some namespace sharing between images, which sounds like a source of hard-to-debug errors.

I saw this comment in the main PR description:

```
set(FOO 0)

add_subdirectory(child_image child_image_build_dir)

print(FOO)
```

> Due to CMake scoping rules, the variable FOO will be '0' after
> 'add_subdirectory' has been processed, even if the subdirectory
> temporarily set FOO to something else.

I believe the reason it doesn't apply to Kconfig variables is because something like this can happen:

- the "main" image build has `CONFIG_WHATEVER=y` in its `.config`, so kconfig.cmake sets `CONFIG_WHATEVER` to `"y"` at CMake level in the application directory scope
- the application directory scope's bindings are used initialize the directory scope of a "child" image when `add_subdirectory()` is called
- the child's `.config` has `# CONFIG_WHATEVER is not set`
- the child's run of `kconfig.cmake` doesn't do anything to `CONFIG_WHATEVER` in the child directory scope, leaving the undesired `CONFIG_WHATEVER=y` from the parent scope in place

This loop would fix that issue, but I'm having trouble understanding why the same situation won't apply to any other variables that are conditionally set by the build system depending on some file, the environment, etc. One example would be calls like this one in openocd.board.cmake:

```
set_ifndef(BOARD_FLASH_RUNNER openocd)
set_ifndef(BOARD_DEBUG_RUNNER openocd)
```

Without a user-provided `BOARD_FLASH_RUNNER`, the application directory scope would set the value to `openocd`, since the boilerplate file pulls `board.cmake` in using `include()`, and the individual `board.cmake` files for boards which support OpenOCD in turn have `include(.../openocd.board.cmake)` lines. The "child" directory scope would inherit this value, so any `set_ifndef(BOARD_FLASH_RUNNER some_other_value)` lines in the child image build would not have the desired effect.

One might argue that multi-image build is currently only targeting multiple images for the same logical target, which should be able to share runners, but the point remains that the behavior is different, and of course that argument won't survive a future where builds targeting different architectures are considered (e.g. ARM runners which expect SWD to be available would of course not work on another image for a RISC-V target, etc).

I would naively guess that this PR means we need to start being very careful to do things like `set(${IMAGE}_variable value)` 100% of the time, and that "raw" calls to `set()` would have to be audited very carefully and generally discouraged.

If that's the case, I'd say we need an `image_set(VAR VALUE)` (and variants for `set_ifndef` and friends) which automatically prepends the ${IMAGE} value to ${VAR}, and would have to mass-convert all of our `set()` calls over.

I don't know CMake well enough, but wonder if a similar argument applies to other properties or cache variables, but I guess so, given lines like this that are showing up in various places now:

```
define_property(GLOBAL PROPERTY ${IMAGE}SOME_PROPERTY
```

It seems like an `image_define_property` conversion might be in order, so we can grep for non-`image_` prefixed versions and make sure they are multi-image safe in each case. Etc.

I'm also concerned about the default IMAGE being the empty string instead of something like `main_image_`, since while the above covers writes, we would also need to be careful about reading the value of any potentially image-specific variable; any code which reads `VAR` when it should be reading `${IMAGE}_VAR` would seem like it was working in single-image builds, then fall apart in strange ways in multi-image builds.

Reply:
> I am concerned about what the need for this loop means about this feature's effect on the build system's namespaces.

The whole thing kind of reminds me of the old Android `include $(CLEAR_VARS)` idiom; https://developer.android.com/ndk/guides/android_mk

Reply:
About introducing 'image_set' wrappers.

Maybe. I considered this but decided not to because I felt the
pattern was easier to understand without a wrapper function than
with.

Note that 'image_set' would also require an 'image_get', and when
CMake does not have return values in it's functions the resulting
code will have a lot of boilerplate.

If we do introduce this we should name it something like
'variant_set', so that we future-proof ourselves to other
variants than images.

Reply:
> I would naively guess that this PR means we need to start being very careful to do things like set(${IMAGE}variable value) 100% of the time, and that "raw" calls to set() would have to be audited very carefully and generally discouraged.

It is luckily not necessary to always use the IMAGE-prefix. It is necessary if and only if;

The variable is image-specific, and
the variable is within scope of multiple images, and
the variable is dereferenced before it is initialized.

Or if it is an image-specific global.

I will improve the documentation to make it clear when and when not to use an IMAGE prefix.

Reply:
> I'm also concerned about the default IMAGE being the empty string instead of something like main_image_

This is for compatibility with existing single-image build scripts, single-image UI (ninja menuconfig), and for keeping the single-image build output clean.

These benefits outweigh the overhead of fixing accidentally introduced build script bugs.